contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,484 @@
+            }
+        );
+        match bond.return_type {
+            ReturnType::Remote => {
+                bond.state = BondState::Sent;
+                BONDS.save(deps.storage, id, &bond)?;
+                msgs.push(CosmosMsg::Custom(NeutronMsg::IbcTransfer {
+                    source_port: source_port.clone(),
+                    source_channel: source_channel.clone(),
+                    token: bond.received.unwrap(),
+                    sender: env.contract.address.to_string(),
+                    receiver: bond.receiver,
+                    timeout_height: RequestPacketTimeoutHeight {
+                        revision_number: None,
+                        revision_height: None,
+                    },
+                    timeout_timestamp: env.block.time.plus_seconds(ibc_timeout).nanos(),
+                    memo: id.to_string(),
+                    fee: query_ibc_fee(deps.as_ref(), LOCAL_DENOM)?,
+                }));
+            }
+            ReturnType::Local => {
+                if let Some(backup) = bond.backup {
+                    BONDS.remove(deps.storage, id);
+                    msgs.push(CosmosMsg::Bank(BankMsg::Send {
+                        to_address: backup,
+                        amount: vec![bond.received.unwrap()],
+                    }));
+                }
+            }
+        }
+    }
+    Ok(response("complete", CONTRACT_NAME, attrs).add_messages(msgs))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn reply(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    msg: Reply,
+) -> ContractResult<Response<NeutronMsg>> {
+    finalize_bond(deps, msg)
+}
+
+pub fn finalize_bond(
+    deps: DepsMut<NeutronQuery>,
+    msg: Reply,
+) -> ContractResult<Response<NeutronMsg>> {
+    match msg.result {
+        cosmwasm_std::SubMsgResult::Ok(res) => {
+            let mut bond = BONDS.load(deps.storage, msg.id)?;
+            bond.state = BondState::Bonded;
+            // get token factory mint event
+            let tf_mint_event = res
+                .events
+                .iter()
+                .find(|x| x.ty == "tf_mint")
+                .ok_or(ContractError::NoTokensMinted)?;
+            // get amount from mint event
+            let coin = Coin::from_str(
+                &tf_mint_event
+                    .attributes
+                    .iter()
+                    .find(|x| x.key == "amount")
+                    .ok_or(ContractError::NoTokensMintedAmountFound)?
+                    .value,
+            )?;
+            bond.received = Some(coin);
+            BONDS.save(deps.storage, msg.id, &bond)?;
+            let attrs = vec![
+                attr("action", "finalize_bond"),
+                attr("id", msg.id.to_string()),
+                attr("state", bond.state.to_string()),
+            ];
+            Ok(response("finalize_bond", CONTRACT_NAME, attrs))
+        }
+        cosmwasm_std::SubMsgResult::Err(_) => unreachable!(),
```

Comment:
I would rather write a comment why this code is unreachable (that's it, because we always use `SubMsg::reply_on_success()`).